### ISSUE TITLE
Run tests in multiple threads

### ISSUE THREAD

**Issue by msokolov** [randomizedtesting/randomizedtesting#252]
I've been using a multithreaded test runner I wrote for a while, and I was trying to see if there was any way to use this with, or maybe contribute it to this library. My runner runs each test case in its own thread; I get very nice speedups with it, and it can also be used to uncover some race conditions by allowing test cases to access shared state in a test class. I was poking around in here and found this comment in RandomizedRunner.runSuite:

    // NOTE: this effectively means we can't run concurrent randomized runners.
    final UncaughtExceptionHandler previous = Thread.getDefaultUncaughtExceptionHandler();

Does this mean it doesn't (and can't) run tests in multiple threads?  I'm not sure where the restriction takes effect: is it global, per-suite, per-class or per-case?

**Comment by dweiss** [comment]
Hi Michael. The reason we don't do multi-threaded parallel test execution is very simple: the JVM doesn't allow full sandboxing of tests and our primary goal is full reproducibility from a single seed. Think of the globals that can be conflicting -- thread management (as you quoted), security policies, system properties, class loading order (and static initializers). For very simple tests that don't interact with the environment parallel execution is great, but anything beyond that is a problem.

A half-baked solution to this is implemented in the ANT runner for randomized testing -- it splits the set of test suites into multiple JVMs (load-balancing tests execution between them). This requires multiple JVMs to start (overhead), but if you have many test suites, the concurrency is quite all right.

In short: it is possible to run tests in parallel within a single JVM, but there are deep-rooted potential more problems with it, so it won't be implemented as part of this project.

**Comment by msokolov** [comment]
I understand the concern about repeatability. It's certainly true that you cannot guarantee a failure will be reproducible in the face of multi-threading in the general case. However I think that's also true of tests which themselves spawn multiple threads, so in some sense this is an impossibly high standard.

In addition to the perf gain for simpler tests, I've found the multithreaded test runner to be useful in cases where I explicitly want to test some supposedly thread-safe class by running tasks that call its methods from multiple threads. In such a case I'm not sure how one could ever guarantee reproducibility though?

Anyway, I still think this could be useful as a selectable option, say by annotating the test class in cases where the test writer explicitly want to enable it.

**Comment by dweiss** [comment]
Tests that fork multiple threads are non-reproducible if they don't have a state. But still -- if something fails you at least get a chance to repeat the same test execution (with the same randomization seed), even if you re-run it multiple times just to see whether the failure is intermittent or permanent for a given seed. Randomized runner tries to ensure every thread forked from the same test has the same initial randomization seed, so it's not really dumb -- it does its best.

bq. I've found the multithreaded test runner to be useful in cases where I explicitly want to test some supposedly thread-safe class by running tasks that call its methods from multiple threads. 

I think a better design here would be to create a single test, then fork a number (randomized!) of threads and pound your business logic from those threads (but still within a single test!). 

Believe me I've been down the route of trying to come up with a clean multithreaded separation of tests within a single JVM, but it quickly hits a showstopper -- be it system property-dependent initialization of something (which many third-party libraries use), a race condition on a singleton somewhere, configured once... you name it.

bq. Anyway, I still think this could be useful as a selectable option, say by annotating the test class in cases where the test writer explicitly want to enable it.

Useful -- maybe. But misleading -- for sure. I bet people would abuse this and then be stuck at how to reproduce a particular problem. If you're looking for difficult-to-reproduce issues even in the current setup, look at Apache Solr tests -- they are notorious for failing on one machine, but not on another...

The better you can isolate a single test run, the easier it is to debug. If you're unhappy about it -- please go ahead and fork the project or create your own runner! I bet somebody wrote something like this already (scan JUnit's mailing list).

**Comment by msokolov** [comment]
Yeah, OK. I *do* have my own runner already. It's not very complicated to write. I'm working in a project that is borrowing LuceneTestCase so trying to find a way to adapt to that. Perhaps we'll just use this runner in some classes and LuceneTestCase in others. Anyway, thanks for the thoughtful response.

**Comment by dweiss** [comment]
You're welcome!
